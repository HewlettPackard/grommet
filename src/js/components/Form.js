// (C) Copyright 2014-2015 Hewlett Packard Enterprise Development LP

import React, { PropTypes } from 'react';
import classnames from 'classnames';

const CLASS_ROOT = 'form';

const Form = props => {
  let { className, compact, fill, pad } = props;
  let classes = classnames(
    CLASS_ROOT,
    className,
    {
      [`${CLASS_ROOT}--compact`]: compact,
      [`${CLASS_ROOT}--fill`]: fill,
      [`${CLASS_ROOT}--pad-${pad}`]: typeof pad === 'string',
      [`${CLASS_ROOT}--pad-horizontal-${pad.horizontal}`]: typeof pad === 'object' && 'horizontal' in pad,
      [`${CLASS_ROOT}--pad-vertical-${pad.vertical}`]: typeof pad === 'object' && 'vertical' in pad
    }
  );

  return (
    <form className={classes} onSubmit={props.onSubmit}>
      {props.children}
    </form>
  );
};

Form.propTypes = {
  compact: PropTypes.bool,
  fill: PropTypes.bool,
  onSubmit: PropTypes.func,
  pad: PropTypes.oneOfType([
    PropTypes.oneOf(['none', 'small', 'medium', 'large']),
    PropTypes.shape({
      horizontal: PropTypes.oneOf(['none', 'small', 'medium', 'large']),
      vertical: PropTypes.oneOf(['none', 'small', 'medium', 'large'])
    })
  ])
};

Form.defaultProps = {
  compact: false,
  fill: false,
  pad: 'none'
};

Form.displayName = 'Form';

export default Form;
