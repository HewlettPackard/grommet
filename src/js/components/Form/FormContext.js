import React, { useEffect, useState } from 'react';

// When not a descendant of a Form, FormContext still provides a basic
// useFormContext that holds the value state.
const useFormContext = (_, valueProp) => {
  const [value, setValue] = useState(valueProp !== undefined ? valueProp : '');
  // use whatever value is passed in, even when it changes
  useEffect(() => setValue(valueProp), [valueProp]);
  return [value, setValue];
};

export const FormContext = React.createContext({ useFormContext });
