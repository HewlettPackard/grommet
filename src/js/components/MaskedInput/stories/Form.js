import React from 'react';
import { storiesOf } from '@storybook/react';

import { Box, Button, Form, FormField, Grommet, MaskedInput } from 'grommet';
import { grommet } from 'grommet/themes';

const Example = () => {
  const [value, setValue] = React.useState({ value: '' });
  const onChange = nextValue => {
    console.log('onChange', nextValue);
    setValue(nextValue);
  };
  return (
    <Grommet theme={grommet}>
      <Box align="center" pad="large">
        <Form
          value={value}
          onChange={onChange}
          onSubmit={({ value: nextValue }) => {
            console.log(nextValue);
            setValue({ value: '' });
          }}
        >
          <FormField name="value" label="url" required>
            <MaskedInput
              name="value"
              mask={[{ fixed: 'https://' }, { regexp: /^.*$/ }]}
            />
          </FormField>
          <Button type="submit" label="submit" />
        </Form>
      </Box>
    </Grommet>
  );
};

storiesOf('MaskedInput', module).add('Form', () => <Example />);
