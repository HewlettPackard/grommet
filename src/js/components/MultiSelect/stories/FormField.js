import React, { useCallback, useState } from 'react';

import { Box, Button, Form, FormField } from 'grommet';
import { MultiSelect } from '../MultiSelect';

const options = [
  'Apple',
  'Orange',
  'Banana',
  'Grape',
  'Melon',
  'Strawberry',
  'Kiwi',
  'Mango',
  'Raspberry',
  'Rhubarb',
];

export const InsideFormField = () => {
  const [value, setValue] = useState({});
  const onChange = useCallback((nextValue) => setValue(nextValue), []);

  return (
    // Uncomment <Grommet> lines when using outside of storybook
    // <Grommet theme={...}>
    <Box align="center" pad="large">
      <Form
        value={value}
        onChange={onChange}
        onSubmit={() => console.log('Submit', value)}
      >
        <FormField label="Label" name="select">
          <MultiSelect
            showSelectedInline
            name="select"
            placeholder="placeholder"
            options={options}
          />
        </FormField>
        <Button type="submit" label="Update" primary />
      </Form>
    </Box>
    // </Grommet>
  );
};

InsideFormField.storyName = 'Inside a FormField';

export default {
  title: 'Input/MultiSelect/Inside a FormField',
};
