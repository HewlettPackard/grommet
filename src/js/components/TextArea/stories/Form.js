import React from 'react';
import { storiesOf } from '@storybook/react';

import { Box, Button, Form, FormField, Grommet, TextArea } from 'grommet';
import { grommet } from 'grommet/themes';

const Example = () => {
  const [value, setValue] = React.useState({ name: '', email: '', value: '' });

  return (
    <Grommet theme={grommet}>
      <Box align="center" pad="large">
        <Form
          value={value}
          onChange={nextValue => setValue(nextValue)}
          onSubmit={() => console.log(value)}
        >
          <FormField name="value" label="value" required>
            <TextArea name="value" />
          </FormField>
          <Button type="submit" label="submit" />
        </Form>
      </Box>
    </Grommet>
  );
};

storiesOf('TextArea', module).add('Form', () => <Example />);
