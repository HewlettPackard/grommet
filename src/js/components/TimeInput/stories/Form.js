import React from 'react';
import { storiesOf } from '@storybook/react';

import { Box, Button, TimeInput, Form, FormField, Grommet } from 'grommet';
import { grommet } from 'grommet/themes';

const Example = () => {
  const [value, setValue] = React.useState({ value: new Date().toISOString() });
  const onChange = nextValue => {
    console.log('onChange', nextValue);
    setValue(nextValue);
  };
  return (
    <Grommet theme={grommet}>
      <Box align="center" pad="large">
        <Form
          value={value}
          onChange={onChange}
          onSubmit={({ value: nextValue }) => console.log(nextValue)}
        >
          <FormField name="value" label="value" required>
            <TimeInput
              name="value"
              format="hh:mm"
              onChange={v => onChange({ value: v })}
            />
          </FormField>
          <Button type="submit" label="submit" />
        </Form>
      </Box>
    </Grommet>
  );
};

storiesOf('TimeInput', module).add('Form', () => <Example />);
