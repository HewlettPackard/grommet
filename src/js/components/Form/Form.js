import React, {
  forwardRef,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import { FormContext } from './FormContext';

const defaultMessages = {
  invalid: 'invalid',
  required: 'required',
};
const defaultValue = {};
const defaultErrors = {};

const Form = forwardRef(
  (
    {
      children,
      errors: errorsProp = defaultErrors,
      messages: messagesProp = defaultMessages,
      onChange,
      onReset,
      onSubmit,
      validate = 'submit',
      value: valueProp = defaultValue,
      ...rest
    },
    ref,
  ) => {
    const [value, setValue] = useState(valueProp);
    useEffect(() => {
      if (valueProp !== defaultValue) setValue(valueProp);
    }, [valueProp]);
    const [messages, setMessages] = useState(messagesProp);
    useEffect(() => setMessages(messagesProp), [messagesProp]);
    const [errors, setErrors] = useState(errorsProp || {});
    useEffect(() => setErrors(errorsProp || {}), [errorsProp]);
    const [touched, setTouched] = useState({});

    const validations = useRef({});

    useEffect(() => {
      if (onChange) onChange(value);
    }, [onChange, value]);

    useEffect(() => {}, [value, errors]);

    const update = useCallback((name, data, error) => {
      setValue(prevValue => {
        const nextValue = { ...prevValue };
        nextValue[name] = data;

        setErrors(prevErrors => {
          const nextErrors = { ...prevErrors };
          if (prevErrors[name]) {
            const nextError =
              error ||
              (validations.current[name] &&
                validations.current[name](data, nextValue));
            if (nextError) {
              nextErrors[name] = nextError;
            } else {
              delete nextErrors[name];
            }
          }
          return nextErrors;
        });

        return nextValue;
      });

      setTouched(prevTouched => {
        const nextTouched = { ...prevTouched };
        nextTouched[name] = true;
        return nextTouched;
      });
    }, []);

    return (
      <form
        ref={ref}
        {...rest}
        onReset={event => {
          setValue(defaultValue);
          setErrors({});
          setTouched({});
          if (onReset) {
            event.persist(); // extract from React's synthetic event pool
            const adjustedEvent = event;
            adjustedEvent.value = defaultValue;
            onReset(adjustedEvent);
          }
        }}
        onSubmit={event => {
          // Don't submit the form via browser form action. We don't want it
          // if the validation fails. And, we assume a javascript action handler
          // otherwise.
          event.preventDefault();
          const nextErrors = { ...errors };
          Object.keys(validations.current).forEach(name => {
            const validation = validations.current[name];
            const error = validation && validation(value[name], value);
            if (error) {
              nextErrors[name] = error;
            } else {
              delete nextErrors[name];
            }
          });
          if (Object.keys(nextErrors).length === 0 && onSubmit) {
            event.persist(); // extract from React's synthetic event pool
            const adjustedEvent = event;
            adjustedEvent.value = value;
            onSubmit(adjustedEvent);
          } else {
            setErrors(nextErrors);
          }
        }}
      >
        <FormContext.Provider
          value={{
            addValidation: (name, validation) => {
              validations.current[name] = validation;
            },
            onBlur:
              validate === 'blur'
                ? name => {
                    if (validations.current[name]) {
                      setErrors(prevErrors => {
                        const nextErrors = { ...prevErrors };
                        const error = validations.current[name](
                          value[name],
                          value,
                        );
                        if (error) {
                          nextErrors[name] = error;
                        } else {
                          delete nextErrors[name];
                        }
                        return nextErrors;
                      });
                    }
                  }
                : undefined,
            errors,
            messages,
            touched,
            update,
            value,
          }}
        >
          {children}
        </FormContext.Provider>
      </form>
    );
  },
);

Form.displayName = 'Form';

let FormDoc;
if (process.env.NODE_ENV !== 'production') {
  FormDoc = require('./doc').doc(Form); // eslint-disable-line global-require
}

const FormWrapper = FormDoc || Form;

export { FormWrapper as Form };
