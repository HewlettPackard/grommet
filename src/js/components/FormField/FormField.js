import React, {
  Children,
  cloneElement,
  forwardRef,
  useContext,
  useEffect,
  useState,
} from 'react';
import styled, { ThemeContext } from 'styled-components';

import { parseMetricToNum } from '../../utils';
import { Box } from '../Box';
import { CheckBox } from '../CheckBox';
import { RadioButtonGroup } from '../RadioButtonGroup';
import { Text } from '../Text';
import { TextInput } from '../TextInput';
import { FormContext } from '../Form/FormContext';

const grommetInputNames = ['TextInput', 'Select', 'MaskedInput', 'TextArea'];
const grommetInputPadNames = ['CheckBox', 'RadioButtonGroup', 'RangeInput'];

const validateField = (required, validate, messages) => (value, data) => {
  let error;
  if (required && (value === undefined || value === '')) {
    error = messages.required;
  } else if (validate) {
    if (Array.isArray(validate)) {
      validate.some(oneValidate => {
        error = validateField(false, oneValidate, messages)(value, data);
        return !!error;
      });
    } else if (typeof validate === 'function') {
      error = validate(value, data);
    } else if (validate.regexp) {
      if (!validate.regexp.test(value)) {
        error = validate.message || messages.invalid;
      }
    }
  }
  return error;
};

const FormFieldBox = styled(Box)`
  ${props => props.theme.formField && props.theme.formField.extend}
`;

const FormField = forwardRef(
  (
    {
      checked,
      children,
      className,
      component,
      disabled,
      error,
      help,
      htmlFor,
      label,
      margin,
      name,
      onBlur,
      onFocus,
      pad,
      required,
      style,
      validate,
      value: valueProp,
      ...rest
    },
    ref,
  ) => {
    const theme = useContext(ThemeContext);
    const context = useContext(FormContext);
    const [value, setValue] = useState(valueProp);
    useEffect(() => setValue(valueProp), [valueProp]);

    useEffect(() => {
      if (
        context &&
        context.value[name] === undefined &&
        (value !== undefined || checked !== undefined)
      ) {
        context.update(
          name,
          value !== undefined ? value : checked,
          undefined,
          true,
        );
      }
    });

    const [focus, setFocus] = useState();

    const renderInput = (formValue, invalid) => {
      const Input = component || TextInput;
      if (Input === CheckBox) {
        return (
          <Input
            name={name}
            label={label}
            checked={
              formValue[name] !== undefined ? formValue[name] : checked || false
            }
            aria-invalid={invalid || undefined}
            {...rest}
          />
        );
      }
      return (
        <Input
          name={name}
          value={
            formValue[name] !== undefined ? formValue[name] : valueProp || ''
          }
          plain
          focusIndicator={false}
          aria-invalid={invalid || undefined}
          {...rest}
        />
      );
    };

    const { formField } = theme;
    const { border } = formField;

    let normalizedError = error;
    // This is here for backwards compatibility. In case the child is a grommet
    // input component, set plain and focusIndicator props, if they aren't
    // already set.
    let wantContentPad =
      component && (component === CheckBox || component === RadioButtonGroup);
    let contents =
      (border &&
        children &&
        Children.map(children, child => {
          if (
            child &&
            child.type &&
            grommetInputPadNames.indexOf(child.type.displayName) !== -1
          ) {
            wantContentPad = true;
          }
          if (
            child &&
            child.type &&
            grommetInputNames.indexOf(child.type.displayName) !== -1 &&
            child.props.plain === undefined &&
            child.props.focusIndicator === undefined
          ) {
            return cloneElement(child, {
              plain: true,
              focusIndicator: false,
            });
          }
          return child;
        })) ||
      children;

    let onFieldBlur;
    if (context) {
      const {
        addValidation,
        errors,
        onBlur: onContextBlur,
        value: formValue,
        messages,
      } = context;
      addValidation(name, validateField(required, validate, messages));
      normalizedError = error || errors[name];
      contents = contents || renderInput(formValue, !!normalizedError);
      if (onContextBlur) {
        onFieldBlur = () => onContextBlur(name);
      }
    }

    const contentProps = pad || wantContentPad ? { ...formField.content } : {};
    if (border.position === 'inner') {
      if (normalizedError && formField.error) {
        contentProps.background = formField.error.background;
      } else if (disabled && formField.disabled) {
        contentProps.background = formField.disabled.background;
      }
    }
    contents = <Box {...contentProps}>{contents}</Box>;

    let borderColor;
    if (focus && !normalizedError) {
      borderColor = 'focus';
    } else if (normalizedError) {
      borderColor = (border && border.error.color) || 'status-critical';
    } else {
      borderColor = (border && border.color) || 'border';
    }
    let abut;
    let abutMargin;
    let outerStyle = style;

    if (border) {
      contents = (
        <Box
          border={
            border.position === 'inner'
              ? {
                  ...border,
                  side: border.side || 'bottom',
                  color: borderColor,
                }
              : undefined
          }
        >
          {contents}
        </Box>
      );

      const mergedMargin = margin || formField.margin;
      abut =
        border.position === 'outer' &&
        (border.side === 'all' ||
          border.side === 'horizontal' ||
          !border.side) &&
        !(
          mergedMargin &&
          ((typeof mergedMargin === 'string' && mergedMargin !== 'none') ||
            (mergedMargin.bottom && mergedMargin.bottom !== 'none') ||
            (mergedMargin.horizontal && mergedMargin.horizontal !== 'none'))
        );
      if (abut) {
        // marginBottom is set to overlap adjacent fields
        abutMargin = { bottom: '-1px' };
        if (margin) {
          abutMargin = margin;
        } else if (border.size) {
          // if the user defines a margin,
          // then the default margin below will be overriden
          abutMargin = {
            bottom: `-${parseMetricToNum(
              theme.global.borderSize[border.size] || border.size,
            )}px`,
          };
        }

        outerStyle = {
          position: focus ? 'relative' : undefined,
          zIndex: focus ? 10 : undefined,
          ...style,
        };
      }
    }

    let outerBackground;
    if (border.position === 'outer') {
      if (normalizedError && formField.error) {
        outerBackground = formField.error.background;
      } else if (disabled && formField.disabled) {
        outerBackground = formField.disabled.background;
      }
    }

    return (
      <FormFieldBox
        ref={ref}
        className={className}
        border={
          border && border.position === 'outer'
            ? { ...border, color: borderColor }
            : undefined
        }
        background={outerBackground}
        margin={abut ? abutMargin : margin || { ...formField.margin }}
        style={outerStyle}
        onFocus={event => {
          setFocus(true);
          if (onFocus) onFocus(event);
        }}
        onBlur={event => {
          setFocus(false);
          if (onFieldBlur) onFieldBlur(event);
          if (onBlur) onBlur(event);
        }}
      >
        {(label && component !== CheckBox) || help ? (
          <>
            {label && component !== CheckBox && (
              <Text as="label" htmlFor={htmlFor} {...formField.label}>
                {label}
              </Text>
            )}
            {help && (
              <Text
                {...formField.help}
                color={formField.help.color[theme.dark ? 'dark' : 'light']}
              >
                {help}
              </Text>
            )}
          </>
        ) : (
          undefined
        )}
        {contents}
        {normalizedError && (
          <Text
            {...formField.error}
            color={formField.error.color[theme.dark ? 'dark' : 'light']}
          >
            {normalizedError}
          </Text>
        )}
      </FormFieldBox>
    );
  },
);

FormField.displayName = 'FormField';

let FormFieldDoc;
if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  FormFieldDoc = require('./doc').doc(FormField);
}
const FormFieldWrapper = FormFieldDoc || FormField;

export { FormFieldWrapper as FormField };
