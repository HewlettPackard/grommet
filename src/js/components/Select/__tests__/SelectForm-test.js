import React from 'react';
import 'jest-styled-components';
import { cleanup, render, fireEvent } from '@testing-library/react';

import { Grommet } from '../..';
import { Select } from '..';
import { Form } from '../../Form';
import { FormField } from '../../FormField';

describe('Select with Form', () => {
  let App;
  beforeEach(() => {
    App = ({ ...props }) => {
      return (
        <Grommet>
          <Form>
            <FormField label="Select">
              <Select {...props} />
            </FormField>
          </Form>
        </Grommet>
      );
    };
  });
  afterEach(cleanup);

  test('renders', () => {
    const { container } = render(<App options={['one', 'two']} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('opens', () => {
    const onOpen = jest.fn();
    const { getByLabelText } = render(
      <App a11yTitle="test-label" options={['one', 'two']} onOpen={onOpen} />,
    );
    fireEvent.click(getByLabelText('test-label'));
    expect(onOpen).toHaveBeenCalled();
  });

  test('select an option', () => {
    const { getByLabelText } = render(
      <App id="test-select" a11yTitle="test-label" options={['one', 'two']} />,
    );
    const select = getByLabelText('test-label');
    fireEvent.click(select);
    fireEvent.click(
      document.getElementById('test-select__drop').querySelector('button'),
    );
    expect(select.value).toEqual('one');
  });
});
