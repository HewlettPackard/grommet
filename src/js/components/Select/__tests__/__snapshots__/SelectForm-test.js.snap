// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select with Form renders 1`] = `
.c12 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #7D4CDB;
  stroke: #7D4CDB;
}

.c12 g {
  fill: inherit;
  stroke: inherit;
}

.c12 *:not([stroke])[fill="none"] {
  stroke-width: 0;
}

.c12 *[stroke*="#"],
.c12 *[STROKE*="#"] {
  stroke: inherit;
  fill: none;
}

.c12 *[fill-rule],
.c12 *[FILL-RULE],
.c12 *[fill*="#"],
.c12 *[FILL*="#"] {
  fill: inherit;
  stroke: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-bottom: 12px;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  border-bottom: solid 1px rgba(0,0,0,0.33);
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-left: 12px;
  margin-right: 12px;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c5 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  outline: none;
  border: none;
  padding: 0;
  text-align: inherit;
}

.c2 {
  margin-left: 12px;
  margin-right: 12px;
  margin-top: 6px;
  margin-bottom: 6px;
  font-size: 18px;
  line-height: 24px;
}

.c9 {
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: none;
  -webkit-appearance: none;
  background: transparent;
  color: inherit;
  padding: 11px;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
  width: 100%;
  outline: none;
  border: none;
}

.c9::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c9::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c9::-moz-placeholder {
  color: #AAAAAA;
}

.c9:-ms-input-placeholder {
  color: #AAAAAA;
}

.c9::-moz-focus-inner {
  border: none;
  outline: none;
}

.c8 {
  position: relative;
  width: 100%;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c10 {
  cursor: pointer;
}

@media only screen and (max-width:768px) {
  .c1 {
    margin-bottom: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c3 {
    border-bottom: solid 1px rgba(0,0,0,0.33);
  }
}

@media only screen and (max-width:768px) {
  .c11 {
    margin-left: 6px;
    margin-right: 6px;
  }
}

<div
  class="c0"
>
  <form>
    <div
      class="c1 "
    >
      <label
        class="c2"
      >
        Select
      </label>
      <div
        class="c3 "
      >
        <div
          class="c4"
        >
          <button
            aria-label="Open Drop"
            class="c5 "
            type="button"
          >
            <div
              class="c6"
            >
              <div
                class="c7"
              >
                <div
                  class="c8"
                >
                  <input
                    autocomplete="off"
                    class="c9 c10"
                    readonly=""
                    tabindex="-1"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="c11"
                style="min-width: auto;"
              >
                <svg
                  aria-label="FormDown"
                  class="c12"
                  viewBox="0 0 24 24"
                >
                  <polyline
                    fill="none"
                    points="18 9 12 15 6 9"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </div>
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
